crates/iota-core/src/congestion_tracker.rs

```diff
@@ -64,9 +73,44 @@ impl CongestionInfo {
         }
     }
 
-    /// Update the highest congestion gas price with the new `gas_price`.
-    fn update_highest_congestion_gas_price(&mut self, gas_price: u64) {
-        self.highest_congestion_gas_price = self.highest_congestion_gas_price.max(gas_price);
+    fn update_hotness(
+        &mut self,
+        new: &CongestionInfo,
+        number_congested_transactions: usize,
+        _number_cleared_transactions: usize,
+    ) {
+        let old_hotness = self.hotness;
+        if number_congested_transactions > 0 {
+            // Update object i's hotness based on the congestion events following the
+            // formula: hotness_i -= SUM(tx)[SUM(i in tx)(hotness_i- gas_price_feedback_tx)]
+            // * LEARNING_RATE / number_congested_transactions.
+            self.hotness -=
+                new.hotness * HOTNESS_ADJUSTMENT_FACTOR / number_congested_transactions as f64;
+            // Hotness decrease is capped by HOTNESS_DECAY_FACTOR.
+            self.hotness = self.hotness.max(old_hotness / HOTNESS_DECAY_FACTOR);
+        } else {
+            self.hotness = old_hotness;
+        }
```

Comment:
The update in the else branch is redundant. I would rewrite the function as follows:

```suggestion
        if number_congested_transactions > 0 {
            // Update object i's hotness based on the congestion events following the
            // formula: hotness_i -= SUM(tx)[SUM(j in tx)(hotness_j - gas_price_feedback)]
            // * HOTNESS_ADJUSTMENT_FACTOR / number_congested_transactions.
            let new_hotness = self.hotness
                - new.hotness * HOTNESS_ADJUSTMENT_FACTOR / number_congested_transactions as f64;
            // Hotness decrease is capped by HOTNESS_DECAY_FACTOR.
            self.hotness = new_hotness.max(self.hotness / HOTNESS_DECAY_FACTOR);
        }
```

Also, note the changes in the comment. We don't have any sum here, so maybe the comment should be updated too.